server/src/main/java/org/candlepin/model/EntitlementCurator.java

```diff
@@ -614,55 +596,59 @@ public int markDependentEntitlementsDirty(Iterable<Entitlement> entitlements, bo
+            int blockSize = Math.min(this.getInBlockSize(), this.getQueryParameterLimit() / 2);
+            DependentEntitlementQuery dependentEntQuery = new DependentEntitlementQuery(
+                this.getEntityManager(), blockSize);
+            Set<String> eids = new HashSet<String>();
+
+            if (entitlementIds != null && entitlementIds.iterator().hasNext()) {
+                // Find all entitlements that are modified by any of the specified entitlements.
+                eids.addAll(dependentEntQuery.findByProvidedProduct(entitlementIds));
+
+                // Distributors require modified product content matched on derived products to flow
+                // downstream in the manifest. Determine if there are any modified entitlements based
+                // on derived products of any distributors and include those as well.
+                //
+                // We do this in a separate query as it gives us a chance to skip the second query
+                // if it does not apply to any of the specified entitlements.
+                //
+                // We do the filter here to avoid any unnecessary work in the
+                // following query and to reduce the number of entitlements to check.
+                Set<String> distributorEntitlements = filterDistributorEntitlementIds(entitlementIds);
+                if (!distributorEntitlements.isEmpty()) {
+                    eids.addAll(dependentEntQuery.findByDerivedProvidedProduct(distributorEntitlements));
+                }
             }
 
-            Query query = this.getEntityManager().createNativeQuery(querySql);
+            // At this point we have all of the entitlements that need to be marked dirty.
+            // Update the affected entitlements, if necessary...
+            return eids.isEmpty() ? 0 : this.markEntitlementsDirty(eids);
+        }
+        return 0;
+    }
 
-            int blockSize = Math.min(this.getInBlockSize(),
-                this.getQueryParameterLimit() / (excludeInput ? 2 : 1));
+    /**
+     * Given a collection of entitlement IDs, determine which belong to a distributor.
+     *
+     * @param entsIdsToFilter the Entitlement IDs to filter
+     * @return all entitlement IDs that belong to a distributor.
+     */
+    public Set<String> filterDistributorEntitlementIds(Iterable<String> entsIdsToFilter) {
+        Set<String> filteredIds = new HashSet<String>();
+        if (entsIdsToFilter != null && entsIdsToFilter.iterator().hasNext()) {
+            String querySql = "SELECT DISTINCT e.id " +
+                "FROM Entitlement e JOIN e.consumer c JOIN c.type t " +
+                "WHERE t.manifest = true AND e.id IN (:entitlement_ids)";
 
-            Set<String> dirtyIds = new HashSet<String>();
+            Query query = this.getEntityManager().createQuery(querySql);
+            int blockSize = Math.min(this.getInBlockSize(), this.getQueryParameterLimit() / 2);
```

Comment:
You don't need to divide the block size by 2 here for the latter case. That was done in the previous query because it had two in-blocks, so we needed to make sure that each block was no more than half the total number of parameters for the entire query.

Reply:
This comment still applies -- we don't need to divide the query limit by two here.